solidity/contracts/bridge/library/FraudVerification.sol

```diff
@@ -0,0 +1,285 @@
+        require(
+            msg.value >= data.parameters.fraudChallengeDepositAmount,
+            "The amount of ETH deposited is too low"
+        );
+
+        bool verificationResult =
+            CheckBitcoinSigs.checkSig(walletPublicKey, sighash, v, r, s);
+        require(verificationResult, "Signature verification failure");
+
+        uint256 challengeKey =
+            uint256(
+                keccak256(abi.encodePacked(walletPublicKey, sighash, v, r, s))
+            );
+
+        FraudChallenge storage fraudChallenge =
+            data.fraudChallenges[challengeKey];
+        require(fraudChallenge.reportedAt == 0, "Fraud already challenged");
+
+        fraudChallenge.challenger = msg.sender;
+        fraudChallenge.ethDepositAmount = msg.value;
+        /* solhint-disable-next-line not-rely-on-time */
+        fraudChallenge.reportedAt = uint32(block.timestamp);
+        fraudChallenge.defended = false;
+
+        // TODO: Consider emitting the event with walletPublicKey in the
+        //       compressed format as it's how we identify wallets in the Bridge.
+        emit FraudChallengeSubmitted(walletPublicKey, sighash, v, r, s);
+    }
+
+    // TODO: description
+    function submitFraudChallengeResponse(
+        Data storage data,
+        bytes memory walletPublicKey,
+        bytes memory preimage,
+        uint8 v,
+        bytes32 r,
+        bytes32 s,
+        bool witness,
+        mapping(uint256 => Bridge.DepositRequest) storage deposits,
+        mapping(uint256 => bool) storage spentMainUTXOs
+    ) external {
+        // TODO: Consider moving `deposits` and `spentMainUTXOs` to this library
+        // and renaming it `UTXO`. It would be responsible for UTXO-related
+        // logic, like UTXO frauds.
+        bytes32 sighash = preimage.hash256();
+
+        bool verificationResult =
+            CheckBitcoinSigs.checkSig(walletPublicKey, sighash, v, r, s);
+        require(verificationResult, "Signature verification failure");
+
+        uint256 challengeKey =
+            uint256(
+                keccak256(abi.encodePacked(walletPublicKey, sighash, v, r, s))
+            );
+
+        require(
+            data.fraudChallenges[challengeKey].reportedAt > 0,
+            "Fraud not challenged"
+        );
+
+        require(
+            !data.fraudChallenges[challengeKey].defended,
+            "Fraud challenge already defended"
+        );
+
+        uint32 sighashType = extractSighashType(preimage);
+        require(sighashType == 1, "Wrong sighash type");
+
+        if (witness) {
+            verifyWitnessPreimage(preimage, deposits, spentMainUTXOs);
+        } else {
+            verifyNonWitnessPreimage(preimage, deposits, spentMainUTXOs);
+        }
+        // If we passed the preimage verification, the wallet has successfully
+        // defended the fraud challenge.
+
+        // TODO: Reward the wallet for successful fraud challenge response.
```

Comment:
This function can be called by anyone. Shouldn't we reward the caller?

Reply:
Done.